lib/source_file.h

```diff
@@ -205,6 +216,8 @@ class SourceInfo final {
     const InputSources *sources = nullptr;
     SourcePosition start = SourcePosition();
     SourcePosition end = SourcePosition();
+    mutable std::vector<Util::Comment *> before;
+    mutable std::vector<Util::Comment *> after;
```

Comment:
Hmm, I don't like mutable here. `mutable` should not be used as a way to circumvent `const`. In C++ it is usually used for members that don't affect logical const-ness of the class, but need to change. E.g. if a class caches some computation result, it might have a mutable member for the cache that is modified by a `const` method. But it the modification is not observable from the outside (with some caveats like timing and thread safety).

Here, the change is clearly observable as you are adding more information to the source info. Instead, what you should do is make these non-`mutable` and when you need to modify a sourceInfo object you create its copy, add comments there and replace the original object with the copy.